modules/axmm/src/aspace.rs

```diff
@@ -0,0 +1,139 @@
+use axhal::paging::{MappingFlags, PageTable};
+use memory_addr::{is_aligned_4k, PhysAddr, VirtAddr, VirtAddrRange};
+
+use crate::paging_err_to_ax_err;
+
+/// The virtual memory address space.
+pub struct AddrSpace {
+    va_range: VirtAddrRange,
+    pt: PageTable,
+}
+
+impl AddrSpace {
+    /// Returns the address space base.
+    pub const fn base(&self) -> VirtAddr {
+        self.va_range.start
+    }
+
+    /// Returns the address space end.
+    pub const fn end(&self) -> VirtAddr {
+        self.va_range.end
+    }
+
+    /// Returns the address space size.
+    pub const fn size(&self) -> usize {
+        self.va_range.size()
+    }
+
+    /// Returns the reference to the inner page table.
+    pub const fn page_table(&self) -> &PageTable {
+        &self.pt
+    }
+
+    /// Returns the root physical address of the inner page table.
+    pub const fn page_table_root(&self) -> PhysAddr {
+        self.pt.root_paddr()
+    }
+
+    /// Checks if the address space contains the given address range.
+    pub const fn contains_range(&self, start: VirtAddr, size: usize) -> bool {
+        self.va_range
+            .contains_range(VirtAddrRange::from_start_size(start, size))
+    }
+
+    /// Creates a new empty address space.
+    pub(crate) fn new_empty(base: VirtAddr, size: usize) -> AxResult<Self> {
+        Ok(Self {
+            va_range: VirtAddrRange::from_start_size(base, size),
+            pt: PageTable::try_new().map_err(|_| AxError::NoMemory)?,
+        })
+    }
+
+    /// Add a new linear mapping.
+    ///
+    /// The mapping is linear, i.e., `start_vaddr` is mapped to `start_paddr`,
+    /// and `start_vaddr + size` is mapped to `start_paddr + size`.
+    ///
+    /// The `flags` parameter indicates the mapping permissions and attributes.
+    ///
+    /// Returns an error if the address range is out of the address space or not
+    /// aligned.
+    pub fn map_linear(
+        &mut self,
+        start_vaddr: VirtAddr,
+        start_paddr: PhysAddr,
+        size: usize,
+        flags: MappingFlags,
+    ) -> AxResult {
+        if !self.contains_range(start_vaddr, size) {
+            return ax_err!(InvalidInput, "address out of range");
+        }
+        if !start_vaddr.is_aligned_4k() || !start_paddr.is_aligned_4k() || !is_aligned_4k(size) {
+            return ax_err!(InvalidInput, "address not aligned");
+        }
+
+        let offset = start_vaddr.as_usize() - start_paddr.as_usize();
+        self.pt
+            .map_region(
+                start_vaddr,
+                |va| PhysAddr::from(va.as_usize() - offset),
+                size,
+                flags,
+                false, // allow_huge
+                false, // flush_tlb_by_page
+            )
+            .map_err(paging_err_to_ax_err)?
+            .flush();
+        Ok(())
+    }
+
+    /// Removes mappings within the specified virtual address range.
+    ///
+    /// Returns an error if the address range is out of the address space or not
```

Comment:
delete "or not"
Returns an error if the address range is out of the address space or with bad alignment.